expression/builtin_time.go

```diff
@@ -332,83 +339,319 @@ type timeDiffFunctionClass struct {
+}
+
+type builtinStringTimeTimeDiffSig struct {
+	baseDurationBuiltinFunc
+}
+
+// evalDuration evals a builtinStringTimeTimeDiffSig.
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_timediff
+func (b *builtinStringTimeTimeDiffSig) evalDuration(row []types.Datum) (d types.Duration, isNull bool, err error) {
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	lhsStr, isNull, err := b.args[0].EvalString(row, sc)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
+	}
+
+	rhs, isNull, err := b.args[1].EvalTime(row, sc)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
+	}
+
+	_, lhs, isDuration, err := convertStringToDuration(lhsStr, b.tp.Decimal, sc)
+	if err != nil {
+		return d, true, errors.Trace(err)
+	}
+
+	if isDuration {
+		return d, true, nil
+	}
+
+	d, isNull, err = calculateTimeDiff(lhs, rhs, sc)
+	return d, isNull, errors.Trace(err)
+}
+
+type builtinStringStringTimeDiffSig struct {
+	baseDurationBuiltinFunc
+}
+
+// evalDuration evals a builtinStringStringTimeDiffSig.
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_timediff
+func (b *builtinStringStringTimeDiffSig) evalDuration(row []types.Datum) (d types.Duration, isNull bool, err error) {
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	lhs, isNull, err := b.args[0].EvalString(row, sc)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
+	}
+
+	rhs, isNull, err := b.args[1].EvalString(row, sc)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
+	}
+
+	fsp := b.tp.Decimal
+	d0, t0, isDuration0, err := convertStringToDuration(lhs, fsp, sc)
+	if err != nil {
+		return d, true, errors.Trace(err)
+	}
+
+	d1, t1, isDuration1, err := convertStringToDuration(rhs, fsp, sc)
+	if err != nil {
+		return d, true, errors.Trace(err)
+	}
+
+	if isDuration0 != isDuration1 {
+		return d, true, nil
+	}
+
+	if isDuration0 {
+		d, isNull, err = calculateDurationTimeDiff(d0, d1, sc)
+	} else {
+		d, isNull, err = calculateTimeDiff(t0, t1, sc)
+	}
+
+	return d, isNull, errors.Trace(err)
+}
+
+type builtinNullTimeDiffSig struct {
+	baseDurationBuiltinFunc
+}
+
+// evalDuration evals a builtinNullTimeDiffSig.
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_timediff
+func (b *builtinNullTimeDiffSig) evalDuration(row []types.Datum) (d types.Duration, isNull bool, err error) {
+	return d, true, nil
+}
+
+func getStrFsp(strArg string, fsp int) int {
+	if n := strings.IndexByte(strArg, '.'); n >= 0 {
+		lenStrFsp := len(strArg[n+1:])
+		if lenStrFsp <= types.MaxFsp {
+			fsp = mathutil.Max(lenStrFsp, fsp)
+		}
+	}
+	return fsp
+}
+
+// convertStringToDuration converts string to duration, it return types.Time because in some case
+// it will converts string to datetime.
+func convertStringToDuration(str string, fsp int, sc *variable.StatementContext) (d types.Duration, t types.Time,
+	isDuration bool, err error) {
+	fsp = getStrFsp(str, fsp)
```

Comment:
is function `getStrFsp()` only used here ? if so, maybe we can eliminate that function ?